File: platforms/android/src/ds/security/Security.java
package ds.security;


import android.util.Log;

import com.born.boss.BNWebViewActivity;
import com.born.boss.MainActivity;

import org.apache.cordova.CallbackContext;
import org.apache.cordova.CordovaArgs;
import org.apache.cordova.CordovaPlugin;
import org.json.JSONException;
import org.json.JSONObject;

import java.io.BufferedReader;
import java.io.DataOutputStream;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.io.UnsupportedEncodingException;
import java.net.HttpURLConnection;
import java.net.URL;
import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;
import java.util.Iterator;
import java.util.Map;
import java.util.Set;
import java.util.SortedMap;
import java.util.TreeMap;

public class Security extends CordovaPlugin {
    public static final String TAG = "Security";
    private static final String BASE_SALT = "a1f538ec9cf94cf2220b34357ac44f5e7a05rt5a";
    private static final String SALT = "c1f538ec9cf94cf5660b94357ac44f5e7a05817a";
    public static final String ERROR_INVALID_PARAMETERS = "参数格式错误";
    public static final String KEY_DEVICE_MODE = "device_mode";
    public static final String KEY_PLATFORM = "platform";
    public static final String KEY_DEVICE_UUID = "device_uuid";
    public static final String KEY_VERSION = "version";
    public static final String KEY_LATITUDE = "latitude";
    public static final String KEY_LONGITUDE = "longitude";
    public static final String KEY_LOGIN = "login";
    public static final String KEY_UID = "uid";
    public static final String KEY_URL = "url";
    public static final String KEY_DB = "db";
    public static final String KEY_SIGN_DATA = "data";
    public static final String KEY_SIGN_TYPE = "type";
    private static final String ENCODING = "UTF-8";
    private static final int TIME = 40 * 1000;
    public static final String SERVER_URL = "http://we-erp.com/MAST/api/app_init";
    private String _httpUrl = "";
    private String _db = "";
    private String _userSecret = "";

    private String LAUNCH_KEY_URL = "launch_url";
    private String LAUNCH_KEY_DB = "launch_db";
    private String LAUNCH_KEY_CID = "launch_cid";
    private String LAUNCH_KEY_CATEGORY = "launch_category";
    private String LAUNCH_KEY_UID = "launch_uid";
    private String LAUNCH_KEY_MOBILE = "launch_mobile";
    private String LAUNCH_KEY_SID = "launch_sid";
    private String LAUNCH_Browser = "launch_browser";

    private boolean hasFetchUserInfo = false;

    @Override
    public boolean execute(String action, CordovaArgs args, CallbackContext callbackContext)
            throws JSONException {

        Log.e(TAG, _userSecret);
        if (action.equals("sign")) {
            this.sign(args,1, callbackContext);
            return true;
        } else if (action.equals("deviceRegister")) {
            this.deviceRegister(args, callbackContext);
            return true;
        } else if (action.equals("signCommon")) {
            this.signCommon(args,0, callbackContext);
            return true;
        } else if (action.equals("deviceInit")) {
            this.deviceInit(args, callbackContext);
            return true;
        }
        else if (action.equals("launch")) {
            this.launch(args, callbackContext);
            return true;
        }

        return false;
    }


    /**
     * 请求签名
     *
     * @param args
     * @param callbackContext
     * @return
     */
    protected boolean sign(CordovaArgs args, Integer type, final CallbackContext callbackContext) {
        final JSONObject params;
        try {
            params = args.getJSONObject(0);
            SortedMap<Object, Object> parameters = new TreeMap<Object, Object>();
            Iterator<String> it = params.keys();
            while (it.hasNext()) {
                String key = (String) it.next();
                String u = params.get(key).toString();
                parameters.put(key, u);
            }
            StringBuffer sb = new StringBuffer();
            Set es = parameters.entrySet();//所有参与传参的参数按照accsii排序（升序）
            Iterator pit = es.iterator();
            while (pit.hasNext()) {
                Map.Entry entry = (Map.Entry) pit.next();
                String k = (String) entry.getKey();
                Object v = entry.getValue();
                if (null != v && !"".equals(v)
                        && !"sign".equals(k) && !"key".equals(k)) {
                    if(sb.toString().length()>0){
                        sb.append("&"+k + "=" + v);
                    }else{
                        sb.append(k + "=" + v);
                    }
                }
            }
            String signData = sb.toString();

            if (signData == null || signData.length() <= 0) {
                callbackContext.error(ERROR_INVALID_PARAMETERS);
            }

            String sign = "";
            sign = md5(signData + _userSecret);

            params.put("sign", sign);
            callbackContext.success(params);


        } catch (Exception e) {
            Log.e(TAG, e.getMessage());
            callbackContext.error(ERROR_INVALID_PARAMETERS);
            return true;
        }
        return true;
    }



    /**
     * 请求签名
     *
     * @param args
     * @param callbackContext
     * @return
     */
    protected boolean signCommon(CordovaArgs args, Integer type, final CallbackContext callbackContext) {
        final JSONObject params;
        try {
            params = args.getJSONObject(0);
            SortedMap<Object, Object> parameters = new TreeMap<Object, Object>();
            Iterator<String> it = params.keys();
            while (it.hasNext()) {
                String key = (String) it.next();
                String u = params.get(key).toString();
                parameters.put(key, u);
            }
            StringBuffer sb = new StringBuffer();
            Set es = parameters.entrySet();//所有参与传参的参数按照accsii排序（升序）
            Iterator pit = es.iterator();
            while (pit.hasNext()) {
                Map.Entry entry = (Map.Entry) pit.next();
                String k = (String) entry.getKey();
                Object v = entry.getValue();
                if (null != v && !"".equals(v)
                        && !"sign".equals(k) && !"key".equals(k)) {

                    if(sb.toString().length()>0){
                        sb.append("&"+k + "=" + v);
                    }else{
                        sb.append(k + "=" + v);
                    }
                }
            }
            String signData = sb.toString();

            if (signData == null || signData.length() <= 0) {
                callbackContext.error(ERROR_INVALID_PARAMETERS);
            }
            String sign = "";
            sign = md5(signData + SALT);

            params.put("sign", sign);
            callbackContext.success(params);


        } catch (Exception e) {
            Log.e(TAG, e.getMessage());
            callbackContext.error(ERROR_INVALID_PARAMETERS);
            return true;
        }
        return true;
    }


    protected boolean deviceInit(CordovaArgs args, final CallbackContext callbackContext) {
        final JSONObject params;
        try {
            params = args.getJSONObject(0);

            String device_mode = params.getString(KEY_DEVICE_MODE);
            String platform = params.getString(KEY_PLATFORM);
            String device_uuid = params.getString(KEY_DEVICE_UUID);
            String version = params.getString(KEY_VERSION);
            String latitude = params.getString(KEY_LATITUDE);
            String longitude = params.getString(KEY_LONGITUDE);
            String login = params.getString(KEY_LOGIN);

            String data = "device_mode=" + device_mode +
                    "&device_uuid=" + device_uuid +
                    "&latitude=" + latitude +
                    "&login=" + login +
                    "&longitude=" + longitude +
                    "&platform=" + platform +
                    "&version=" + version;
            String sign = md5(data + BASE_SALT);

            String sendData = "<xml><platform>" + platform + "</platform>" +
                    "<longitude>" + longitude + "</longitude>" +
                    "<device_uuid>" + device_uuid + "</device_uuid>" +
                    "<latitude>" + latitude + "</latitude>" +
                    "<device_mode>" + device_mode + "</device_mode>" +
                    "<version>" + version + "</version>" +
                    "<login>" + login + "</login>" +
                    "<sign>" + sign + "</sign>" +
                    "</xml>";

            JSONObject response = sendPostRequest(SERVER_URL, sendData);
            if (response == null) {
                callbackContext.error(ERROR_INVALID_PARAMETERS);
            } else {
                int errcode = response.getInt("errcode");
                if (errcode == 0) {
                    JSONObject rdata = response.getJSONObject("data");

                    String url = rdata.getString("url");
                    String db = rdata.getString("db");

                    this._httpUrl = url;
                    this._db = db;
                    callbackContext.success(rdata);
                } else {
                    String errmsg = response.getString("errmsg");
                    callbackContext.error(errmsg);
                }
            }
        } catch (Exception e) {
            Log.e(TAG, e.getMessage());
            callbackContext.error(ERROR_INVALID_PARAMETERS);
            return true;
        }
        return true;
    }

    protected boolean deviceRegister(CordovaArgs args, final CallbackContext callbackContext) {
        final JSONObject params;
        try {
            params = args.getJSONObject(0);

            String device_mode = params.getString(KEY_DEVICE_MODE);
            String platform = params.getString(KEY_PLATFORM);
            String device_uuid = params.getString(KEY_DEVICE_UUID);
            String version = params.getString(KEY_VERSION);
            String latitude = params.getString(KEY_LATITUDE);
            String longitude = params.getString(KEY_LONGITUDE);
            String login = params.getString(KEY_LOGIN);
            String uid = params.getString(KEY_UID);
            String url = params.getString(KEY_URL);
            String db = params.getString(KEY_DB);
            if (url != null && url.length() > 0) {
                this._httpUrl = url;
            }
            if (db != null && db.length() > 0) {
                this._db = db;
            }

            String data = "device_mode=" + device_mode +
                    "&device_uuid=" + device_uuid +
                    "&latitude=" + latitude +
                    "&login=" + login +
                    "&longitude=" + longitude +
                    "&platform=" + platform +
                    "&uid=" + uid +
                    "&version=" + version;
            String sign = md5(data + SALT);

            String sendData = "<xml><platform>" + platform + "</platform>" +
                    "<longitude>" + longitude + "</longitude>" +
                    "<device_uuid>" + device_uuid + "</device_uuid>" +
                    "<latitude>" + latitude + "</latitude>" +
                    "<device_mode>" + device_mode + "</device_mode>" +
                    "<version>" + version + "</version>" +
                    "<login>" + login + "</login>" +
                    "<uid>" + uid + "</uid>" +
                    "<sign>" + sign + "</sign>" +
                    "</xml>";


            String apiUrl = this._httpUrl + "/" + this._db + "/ds/device";
            JSONObject response = sendPostRequest(apiUrl, sendData);
            if (response == null) {
                callbackContext.error(ERROR_INVALID_PARAMETERS);
            } else {
                int errcode = response.getInt("errcode");
                if (errcode == 0) {
                    JSONObject rdata = response.getJSONObject("data");
                    String secret = rdata.getString("secret");
                    int id = rdata.getInt("id");
                    _userSecret = secret;
                    callbackContext.success(id);
                } else {
                    String errmsg = response.getString("errmsg");
                    callbackContext.error(errmsg);
                }
            }
        } catch (Exception e) {
            Log.e(TAG, e.getMessage());
            callbackContext.error(ERROR_INVALID_PARAMETERS);
            return true;
        }
        return true;
    }

    /**
     * 发送请求
     *
     * @param service 服务地址
     * @param data    数据
     */
    private JSONObject sendPostRequest(String service, String data) {
        HttpURLConnection conn = null;
        InputStream input = null;
        StringBuffer sBuffer = new StringBuffer();
        String line = null;
        BufferedReader bReader = null;
        try {
            URL url = new URL(service);
            conn = (HttpURLConnection) url.openConnection();
            conn.setConnectTimeout(TIME);
            conn.setReadTimeout(TIME);
            conn.setDoInput(true);
            conn.setDoOutput(true);
            conn.setUseCaches(false);
            conn.setRequestMethod("POST");
            conn.setRequestProperty("Connection", "close");
            conn.setRequestProperty("Charset", ENCODING);
            conn.setRequestProperty("Content-Length", String.valueOf(data.length()));
            conn.setRequestProperty("Content-Type", "text/*;charset=utf-8");
            DataOutputStream outStream = new DataOutputStream(conn.getOutputStream());
            outStream.write(data.getBytes());
            outStream.flush();
            outStream.close();
            if (conn == null) {
                return null;
            }
            int responseCode = conn.getResponseCode();
            if (responseCode == 200) {
                input = conn.getInputStream();
                if (input != null) {
                    bReader = new BufferedReader(new InputStreamReader(input));
                    while ((line = bReader.readLine()) != null) {
                        sBuffer.append(line);
                    }
                    return new JSONObject(sBuffer.toString());
                }
            }
        } catch (Exception e) {
            Log.e(TAG, e.getMessage());
        }
        return null;
    }

    /**
     * MD5加密
     *
     * @param string
     * @return
     */
    public static String md5(String string) {
        byte[] hash;
        try {
            hash = MessageDigest.getInstance("MD5").digest(string.getBytes("UTF-8"));
        } catch (NoSuchAlgorithmException e) {
            throw new RuntimeException("Huh, MD5 should be supported?", e);
        } catch (UnsupportedEncodingException e) {
            throw new RuntimeException("Huh, UTF-8 should be supported?", e);
        }
        StringBuilder hex = new StringBuilder(hash.length * 2);
        for (byte b : hash) {
            if ((b & 0xFF) < 0x10) hex.append("0");
            hex.append(Integer.toHexString(b & 0xFF));
        }
        return hex.toString();
    }

    protected boolean launch(CordovaArgs args, final CallbackContext callbackContext) {
        final JSONObject params;
        try {
            params = args.getJSONObject(0);

            String launchUrl = params.getString(LAUNCH_KEY_URL);
            String launchDb = params.getString(LAUNCH_KEY_DB);
            String launchCid = params.getString(LAUNCH_KEY_CID);
            String launchCategory = params.getString(LAUNCH_KEY_CATEGORY);
            String launchUid = params.getString(LAUNCH_KEY_UID);
            String launchMobile = params.getString(LAUNCH_KEY_MOBILE);
            final String launchSid = params.getString(LAUNCH_KEY_SID);

            callbackContext.success("");

            if (launchCategory.equals("browser")) {
                final String url = params.getString(LAUNCH_Browser);

                ((MainActivity) cordova.getActivity()).getHandler().postDelayed(new Runnable() {
                    @Override
                    public void run() {
                        BNWebViewActivity.startAct(cordova.getActivity(),url);
                        //HistoryActivity.startAct(cordova.getActivity());
                    }
                }, 0);
            }
        } catch (Exception e) {
            Log.e(TAG, e.getMessage());
            callbackContext.error(ERROR_INVALID_PARAMETERS);
            return true;
        }
        return true;
    }
}

File: plugins/cordova-plugin-security/src/ios/Security.m
/********* Security.m Cordova Plugin Implementation *******/

#import <Cordova/CDV.h>
#import <CommonCrypto/CommonDigest.h>
#import "HistoryOperateViewController.h"
#import "ProductProjectMainController.h"
#import "GuadanViewController.h"
#import "HistoryOperateViewController.h"
#import "ProjectViewController.h"
#import "StaffViewController.h"
#import "PanDianViewController.h"

@interface Security : CDVPlugin {
    // Member variables go here.
}


@end


static NSString* TAG = @"Security";
static NSString* BASE_SALT = @"a1f538ec9cf94cf2220b34357ac44f5e7a05rt5a";
static NSString* SALT = @"c1f538ec9cf94cf5660b94357ac44f5e7a05817a";
static NSString* ERROR_INVALID_PARAMETERS = @"参数格式错误";
static NSString* KEY_MODE = @"device_mode";
static NSString* KEY_PLATFORM = @"platform";
static NSString* KEY_UUID = @"device_uuid";
static NSString* KEY_VERSION = @"version";
static NSString* KEY_LATITUDE = @"latitude";
static NSString* KEY_LONGITUDE = @"longitude";
static NSString* KEY_LOGIN = @"login";
static NSString* KEY_UID = @"uid";
static NSString* KEY_SIGN_DATA = @"data";
static NSString* KEY_SIGN_TYPE = @"type";
static NSString* ENCODING = @"UTF-8";
#ifdef DEBUG
static NSString* SERVER_URL = @"http://devtop.we-erp.com/MAST/api/app_init";
#else
static NSString* SERVER_URL = @"http://we-erp.com/MAST/api/app_init";
#endif
static NSString* KEY_URL = @"url";
static NSString* KEY_DB = @"db";
static NSString* _httpUrl = @"";
static NSString* _db = @"";
static NSString* _userSecret = @"";

static NSString* LAUNCH_KEY_URL = @"launch_url";
static NSString* LAUNCH_KEY_DB = @"launch_db";
static NSString* LAUNCH_KEY_CID = @"launch_cid";
static NSString* LAUNCH_KEY_CATEGORY = @"launch_category";
static NSString* LAUNCH_KEY_UID = @"launch_uid";
static NSString* LAUNCH_KEY_MOBILE = @"launch_mobile";
static NSString* LAUNCH_KEY_SID = @"launch_sid";
@implementation Security



- (void)launch:(CDVInvokedUrlCommand*)command
{
    [self.commandDelegate runInBackground:^{
        
        
        CDVPluginResult* pluginResult = nil;
        NSDictionary* params = [command.arguments objectAtIndex:0];
        
        NSString* launchUrl = params[LAUNCH_KEY_URL];
        NSString* launchDb = params[LAUNCH_KEY_DB];
        NSString* launchCid = params[LAUNCH_KEY_CID];
        NSString* launchCategory = params[LAUNCH_KEY_CATEGORY];
        NSString* launchUid = params[LAUNCH_KEY_UID];
        NSString* launchMobile = params[LAUNCH_KEY_MOBILE];
        NSString* launchSid = params[LAUNCH_KEY_SID];
        
        //会员管理  member
        //收银  pos
        //挂单  resting
        //业绩考核  follow
        //仓库 stock
        //员工 employee
        
        pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK];
        [self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
        
        dispatch_async(dispatch_get_main_queue(), ^{
            
            PersonalProfile *profile = [PersonalProfile currentProfile];
            if ( ![profile.userName isEqualToString:launchMobile] )
            {
                [PersonalProfile deleteProfile];
                profile = [PersonalProfile currentProfile];
            }
            
            if ( ![profile.isLogin boolValue] )
            {
                PersonalProfile *profile = [[PersonalProfile alloc] init];
                profile.baseUrl = launchUrl;
                profile.sql = launchDb;
                profile.userID = @([launchUid integerValue]);
                profile.userName = launchMobile;
                profile.deviceString = launchCid;
                
                [profile save];
                
                [ICKeyChainManager setCurrentServiceName:[profile.userID stringValue]];
                [BSCoreDataManager setCurrentUserName:[profile.userID stringValue]];
                
                BSLoginRequestStep3 *request = [[BSLoginRequestStep3 alloc] init];
                request.isLogin = TRUE;
                [request execute];
            }
            
            profile.bshopId = @(launchSid.integerValue);
            profile.homeSelectedShopID = @(launchSid.integerValue);
            [profile save];
            
            if ( [launchCategory isEqualToString:@"member"] )
            {
                CDStore *store = [[BSCoreDataManager currentManager] uniqueEntityForName:@"CDStore" withValue:launchSid forKey:@"storeID"];
                MemberViewController *viewController = [[MemberViewController alloc] initWithStore:store];
                [(UINavigationController*)[UIApplication sharedApplication].keyWindow.rootViewController pushViewController:viewController animated:YES];
                
                
                //                HistoryOperateViewController *historyPosVC = [[HistoryOperateViewController alloc] init];
                //                [(UINavigationController*)[UIApplication sharedApplication].keyWindow.rootViewController pushViewController:historyPosVC animated:YES];
                
            }
            else if ([launchCategory isEqualToString:@"hr"])
            {
                StaffViewController *staffVC = [[StaffViewController alloc] initWithNibName:NIBCT(@"StaffViewController") bundle:nil];
                staffVC.isOnlyShop = true;
                NSNumber *storeId = [PersonalProfile currentProfile].homeSelectedShopID;
                staffVC.shop = [[BSCoreDataManager currentManager] findEntity:@"CDStore" withValue:storeId forKey:@"storeID"];
                [(UINavigationController*)[UIApplication sharedApplication].keyWindow.rootViewController pushViewController:staffVC animated:YES];
            }
            else if ([launchCategory isEqualToString:@"product"])
            {
                ProductProjectMainController *procuctVC = [[ProductProjectMainController alloc] init];
                [(UINavigationController*)[UIApplication sharedApplication].keyWindow.rootViewController pushViewController:procuctVC animated:YES];
            }
            else if ([launchCategory isEqualToString:@"pos"])
            {
                ProductProjectMainController *projectMainVC = [[ProductProjectMainController alloc] init];
                projectMainVC.controllerType = ProductControllerType_Sale;
                [(UINavigationController*)[UIApplication sharedApplication].keyWindow.rootViewController pushViewController:projectMainVC animated:YES];
            }
            else if ([launchCategory isEqualToString:@"order"])
            {
                GuadanViewController *guaDanVC = [[GuadanViewController alloc] init];
                [(UINavigationController*)[UIApplication sharedApplication].keyWindow.rootViewController pushViewController:guaDanVC animated:YES];
            }
            else if ([launchCategory isEqualToString:@"historypos"])
            {
                HistoryOperateViewController *historyPosVC = [[HistoryOperateViewController alloc] init];
                [(UINavigationController*)[UIApplication sharedApplication].keyWindow.rootViewController pushViewController:historyPosVC animated:YES];
            }
            else if ([launchCategory isEqualToString:@"inventory"])
            {
                PanDianViewController *panDianVC = [[PanDianViewController alloc] initWithNibName:NIBCT(@"PanDianViewController") bundle:nil];
                [(UINavigationController*)[UIApplication sharedApplication].keyWindow.rootViewController pushViewController:panDianVC animated:YES];
            }
        });
    }];
}

- (void)sign:(CDVInvokedUrlCommand*)command
{
    [self.commandDelegate runInBackground:^{
        CDVPluginResult* pluginResult = nil;
        NSDictionary* params = [command.arguments objectAtIndex:0];
        
        NSArray* allkeys = [params.allKeys sortedArrayUsingComparator:^NSComparisonResult(id  _Nonnull obj1, id  _Nonnull obj2) {
            return [obj1 compare:obj2];
        }];
        
        NSMutableString* signData = [[NSMutableString alloc] init];
        [allkeys enumerateObjectsUsingBlock:^(id  _Nonnull key, NSUInteger idx, BOOL * _Nonnull stop) {
            NSString* value = params[key];
            [signData appendFormat:@"%@=%@",key,value];
            if ( idx != allkeys.count - 1 )
            {
                [signData appendString:@"&"];
            }
        }];
        
        if ( signData.length == 0 || _userSecret.length == 0)
        {
            pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR messageAsString:ERROR_INVALID_PARAMETERS];
        }
        else
        {
            NSString* sign = [self md5Hash:[NSString stringWithFormat:@"%@%@",signData,_userSecret]];
            NSMutableDictionary* dict = [NSMutableDictionary dictionaryWithDictionary:params];
            dict[@"sign"] = sign;
            pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK messageAsDictionary:dict];
        }
        
        [self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
    }];
}


- (void)signCommon:(CDVInvokedUrlCommand*)command
{
    [self.commandDelegate runInBackground:^{
        CDVPluginResult* pluginResult = nil;
        NSDictionary* params = [command.arguments objectAtIndex:0];
        
        NSArray* allkeys = [params.allKeys sortedArrayUsingComparator:^NSComparisonResult(id  _Nonnull obj1, id  _Nonnull obj2) {
            return [obj1 compare:obj2];
        }];
        
        NSMutableString* signData = [[NSMutableString alloc] init];
        [allkeys enumerateObjectsUsingBlock:^(id  _Nonnull key, NSUInteger idx, BOOL * _Nonnull stop) {
            NSString* value = params[key];
            [signData appendFormat:@"%@=%@",key,value];
            if ( idx != allkeys.count - 1 )
            {
                [signData appendString:@"&"];
            }
        }];
        
        NSString* sign = [self md5Hash:[NSString stringWithFormat:@"%@%@",signData,SALT]];
        NSMutableDictionary* dict = [NSMutableDictionary dictionaryWithDictionary:params];
        dict[@"sign"] = sign;
        pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK messageAsDictionary:dict];
        
        [self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
    }];
}


- (void)deviceInit:(CDVInvokedUrlCommand*)command
{
    [self.commandDelegate runInBackground:^{
        CDVPluginResult* pluginResult = nil;
        NSDictionary* params = [command.arguments objectAtIndex:0];
        
        NSString* device_mode = params[KEY_MODE];
        NSString* platform = params[KEY_PLATFORM];
        NSString* device_uuid = params[KEY_UUID];
        NSString* version = params[KEY_VERSION];
        NSString* latitude = params[KEY_LATITUDE];
        NSString* longitude = params[KEY_LONGITUDE];
        NSString* login = params[KEY_LOGIN];
        
        
        NSString* data = [NSString stringWithFormat:@"device_mode=%@&device_uuid=%@&latitude=%@&login=%@&longitude=%@&platform=%@&version=%@",device_mode,device_uuid,latitude,login,longitude,platform,version];
        
        NSString* sign = [self md5Hash:[NSString stringWithFormat:@"%@%@",data,BASE_SALT]];
        
        NSString* sendData = [NSString stringWithFormat:@"<xml><platform>%@</platform><longitude>%@</longitude><latitude>%@</latitude><device_uuid>%@</device_uuid><device_mode>%@</device_mode><version>%@</version><login>%@</login><sign>%@</sign></xml>",platform,longitude,latitude,device_uuid,device_mode,version,login,sign];
        
        
        NSMutableURLRequest* rq = [NSMutableURLRequest requestWithURL: [NSURL URLWithString:SERVER_URL] cachePolicy:NSURLRequestReloadIgnoringCacheData timeoutInterval:30];
        [rq setHTTPMethod: @"POST"];
        [rq setValue:[NSString stringWithFormat:@"%ld",[sendData length]] forHTTPHeaderField:@"Content-Length"];
        [rq setValue:@"text/*;charset=utf-8" forHTTPHeaderField:@"Content-Type"];
        [rq setHTTPBody:[sendData dataUsingEncoding:NSUTF8StringEncoding]];
        
        NSError *error = nil;
        NSData *responseData = [NSURLConnection sendSynchronousRequest:rq returningResponse:nil error:&error];
        if ( responseData == nil )
        {
            pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR messageAsString:ERROR_INVALID_PARAMETERS
                            ];
        }
        else
        {
            NSString *responseString = [[NSString alloc] initWithData:responseData encoding:NSUTF8StringEncoding];
            NSLog(@"response: %@", responseString);
            
            NSDictionary* params = [NSJSONSerialization JSONObjectWithData:responseData options: NSJSONReadingMutableContainers error:nil];
            NSInteger errcode = [params[@"errcode"] integerValue];
            if ( errcode == 0 )
            {
                NSDictionary* rdata = params[@"data"];
                _httpUrl = rdata[@"url"];
                _db = rdata[@"db"];
                pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK messageAsDictionary:rdata];
                
            }
            else
            {
                NSString* errmsg = params[@"errmsg"];
                pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR messageAsString:errmsg
                                ];
            }
        }
        
        [self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
    }];
}




- (void)deviceRegister:(CDVInvokedUrlCommand*)command
{
    [self.commandDelegate runInBackground:^{
        CDVPluginResult* pluginResult = nil;
        NSDictionary* params = [command.arguments objectAtIndex:0];
        
        NSString* device_mode = params[KEY_MODE];
        NSString* platform = params[KEY_PLATFORM];
        NSString* device_uuid = params[KEY_UUID];
        NSString* version = params[KEY_VERSION];
        NSString* latitude = params[KEY_LATITUDE];
        NSString* longitude = params[KEY_LONGITUDE];
        NSString* uid = params[KEY_UID];
        NSString* login = params[KEY_LOGIN];
        NSString* url = params[KEY_URL];
        NSString* db = params[KEY_DB];
        
        if ( url.length > 0 ){
            _httpUrl = url;
        }
        if ( db.length > 0 ){
            _db = db;
        }
        
        NSString* data = [NSString stringWithFormat:@"device_mode=%@&device_uuid=%@&latitude=%@&login=%@&longitude=%@&platform=%@&uid=%@&version=%@",device_mode,device_uuid,latitude,login,longitude,platform,uid,version];
        
        NSString* sign = [self md5Hash:[NSString stringWithFormat:@"%@%@",data,SALT]];
        
        NSString* sendData = [NSString stringWithFormat:@"<xml><platform>%@</platform><longitude>%@</longitude><latitude>%@</latitude><device_uuid>%@</device_uuid><device_mode>%@</device_mode><version>%@</version><uid>%@</uid><login>%@</login><sign>%@</sign></xml>",platform,longitude,latitude,device_uuid,device_mode,version,uid,login,sign];
        
        NSString* apiUrl = [NSString stringWithFormat:@"%@/%@/ds/device",_httpUrl,_db];
        
        NSMutableURLRequest* rq = [NSMutableURLRequest requestWithURL: [NSURL URLWithString:apiUrl] cachePolicy:NSURLRequestReloadIgnoringCacheData timeoutInterval:30];
        [rq setHTTPMethod: @"POST"];
        [rq setValue:[NSString stringWithFormat:@"%ld",[sendData length]] forHTTPHeaderField:@"Content-Length"];
        [rq setValue:@"text/*;charset=utf-8" forHTTPHeaderField:@"Content-Type"];
        [rq setHTTPBody:[sendData dataUsingEncoding:NSUTF8StringEncoding]];
        
        NSError *error = nil;
        NSData *responseData = [NSURLConnection sendSynchronousRequest:rq returningResponse:nil error:&error];
        if ( responseData == nil )
        {
            pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR messageAsString:ERROR_INVALID_PARAMETERS
                            ];
        }
        else
        {
            NSString *responseString = [[NSString alloc] initWithData:responseData encoding:NSUTF8StringEncoding];
            NSLog(@"response: %@", responseString);
            
            NSDictionary* params = [NSJSONSerialization JSONObjectWithData:responseData options: NSJSONReadingMutableContainers error:nil];
            NSInteger errcode = [params[@"errcode"] integerValue];
            if ( errcode == 0 )
            {
                NSDictionary* rdata = params[@"data"];
                NSString* secret = rdata[@"secret"];
                NSInteger cid = [rdata[@"id"] integerValue];
                _userSecret = secret;
                pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK messageAsInt:(int)cid];
            }
            else
            {
                NSString* errmsg = params[@"errmsg"];
                pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR messageAsString:errmsg
                                ];
            }
        }
        
        [self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
    }];
}

- (NSString*)md5Hash:(NSString*)signString
{
    NSData* sign = [signString dataUsingEncoding:NSUTF8StringEncoding];
    
    unsigned char result[CC_MD5_DIGEST_LENGTH];
    CC_MD5([sign bytes], (unsigned int)[sign length], result);
    
    return [NSString stringWithFormat:
            @"%02x%02x%02x%02x%02x%02x%02x%02x%02x%02x%02x%02x%02x%02x%02x%02x",
            result[0], result[1], result[2], result[3], result[4], result[5], result[6], result[7],
            result[8], result[9], result[10], result[11], result[12], result[13], result[14], result[15]
            ];
}

@end
